engine/verification/test/assignerfetcher_test.go

```diff
@@ -0,0 +1,304 @@
+package test
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/stretchr/testify/assert"
+	testifymock "github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/engine/testutil"
+	testmock "github.com/onflow/flow-go/engine/testutil/mock"
+	"github.com/onflow/flow-go/engine/verification/assigner"
+	"github.com/onflow/flow-go/engine/verification/assigner/blockconsumer"
+	"github.com/onflow/flow-go/engine/verification/fetcher/chunkconsumer"
+	mockfetcher "github.com/onflow/flow-go/engine/verification/fetcher/mock"
+	"github.com/onflow/flow-go/engine/verification/utils"
+	"github.com/onflow/flow-go/model/chunks"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/module/mock"
+	"github.com/onflow/flow-go/module/trace"
+	bstorage "github.com/onflow/flow-go/storage/badger"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// TestAssignerFetcherPipeline evaluates behavior of the pipeline of
+// block reader -> block consumer -> assigner engine -> chunks queue -> chunks consumer -> chunks processor (i.e., fetcher engine)
+// block reader receives (container) finalized blocks that contain execution receipts preceding (reference) blocks.
+// some receipts have duplicate results.
+// - in a staked verification node:
+// -- for each distinct result assigner engine receives, it does the chunk assignment and passes the
+// chunk locators of assigned chunks to chunk queue, which in turn delivers to chunks processor though the chunks consumer.
+// - in an unstaked verification node:
+// -- execution results are discarded.
+// - it does a correct resource clean up of the pipeline after handling all incoming receipts
+func TestAssignerFetcherPipeline(t *testing.T) {
+	testcases := []struct {
```

Comment:
The tests are great. Thanks for putting the effort to make the tests clean and reusable. 

One comment is that the covered cases are similar to [the tests cases for the engines](https://github.com/onflow/flow-go/blob/master/engine/verification/assigner/engine_test.go#L125-L144)

My opinion is to having tests covering unique cases from different aspects. 

For Engine level test cases, we'd better having test cases testing its _behavior_.  I've checked the engine test cases, they had good coverage, but the assertions aren't enough. 

I would suggest to update the assertion on the `chunkQueues.StoreChunkLocator` call, to assert the arguments passed into the calls. 

We could simply mock the method to store all the arguments, and check it contains the exact chunk locators we want to store. Just like [what you did here](https://github.com/onflow/flow-go/pull/573/files#diff-c245a6af78175672139a652df0bb3f224d1f342a84400f83eb8587e7e1672213R265).

And we can do that in a separate PR, not necessarily in this PR.

Reply:
Thanks, Leo, sure, I created this issue for it, and going to refactor the test right after my current ongoing (fetcher engine/requester) issues.